### DISCUSSION TITLE
EntryHash Batch Error

### DISCUSSION THREAD

**Discussion by kylieritchie** [moov-io/ach#1077] [Q&A] [answered]
Working with the AWS Lambda go function and I'm able to generate a NACHA file, however, when I use the [Moov ACH File Parser](https://oss.moov.io/ach/) and try to convert the NACHA file from Lambda back to JSON, I get a batch error. 

> unable to parse ach file: line:8 record:Batches *ach.BatchError batch #1 (PPD) EntryHash calculated 63399382 is out-of-balance with batch control 633993843.

I've tried to make the entry hash in the JSON I give to Lambda both 63399382 and 633993843, but Lambda always gives back 633993843 in the NACHA file. The only way I can get the Moov File ACH File Parser to convert the NACHA file to JSON is if I manually change 633993843 to 6339938200. Any idea what's going on here and how I can find a solution?

It seems like Lambda go is not using the first 8-digits of routing number. When I manually add together the first 8 digits of the routing numbers I get what the Moov ACH file parser is looking for which is 63399382(00). When I use every digit of the routing (sometimes number is 9 digits) I get what Lambda calculated: 633993843(0).

This is code I found in iatBatch.go

```
// calculateEntryHash This field is prepared by hashing the 8-digit Routing Number in each entry.
// The Entry Hash provides a check against inadvertent alteration of data
func (iatBatch *IATBatch) calculateEntryHash() string {
	hash := 0
	for _, entry := range iatBatch.Entries {

		entryRDFI, _ := strconv.Atoi(entry.RDFIIdentification)

		hash = hash + entryRDFI
	}
	return iatBatch.numericField(hash, 10)
}
```

**Comment by adamdecaf** [comment]
What version of the ach library are you using? Are you bundling the Go executable or compiling into another language? 

Thanks for digging in and finding out the issue is with using only part of the routing number. 

Also, can you share the routing number in question? I wonder if it starts with a zero which causes the parsing to fail? 

Edit: https://go.dev/play/p/nfysvIvgFIq

**Comment by kylieritchie** [comment]
Using the latest version and bundling the Go executable. Routing numbers are: 

101201863
274973183
111900659
101918101
044000037

The sum using all nine digits of routing is 633993843 (what lambda returns).
The sum using first 8 digits of routing is 63399382  (what the EntryHash using Moov ACH File Parser calculated).

**Comment by adamdecaf** [comment, marked as answer]
@kylieritchie I think I've found a fix. There was an issue with `calculateEntryHash()` using 9 digit routing numbers, but the Nacha guidelines state it must be the first 8 digits. I also made a change to properly handle overflow for IAT batches.

Changes: https://github.com/moov-io/ach/pull/1078

Can you try out this branch? You can `go get github.com/moov-io/ach@fix-discussion-1077` inside a project.

**Reply by kylieritchie** [reply]
It works now! Thank you!

**Reply by adamdecaf** [reply]
I'm releasing v1.19.3 with this fix.
